src/systems/systems/src/hydraulic/linear_actuator.rs

```diff
@@ -0,0 +1,707 @@
+        }
+    }
+
+    fn close_control_valves(&mut self) {
+        if self.mode != LinearActuatorMode::ClosedValves {
+            self.closed_valves_reference_position = self.position_normalized;
+            self.mode = LinearActuatorMode::ClosedValves;
+        }
+    }
+
+    fn open_control_valves_for_position_control(&mut self) {
+        self.mode = LinearActuatorMode::PositionControl;
+    }
+
+    fn set_position_target(&mut self, target_position: Ratio) {
+        self.opened_valves_target_position = target_position;
+    }
+
+    fn compute_control_force(&mut self, hydraulic_pressure: Pressure) {
+        let position_error = self.opened_valves_target_position - self.position_normalized;
+
+        let mut open_loop_flow_target = 0.;
+        if position_error >= Ratio::new::<ratio>(0.001) {
+            open_loop_flow_target = self.max_flow.get::<gallon_per_second>();
+        } else if position_error <= Ratio::new::<ratio>(-0.001) {
+            open_loop_flow_target = self.min_flow.get::<gallon_per_second>();
+        }
+
+        let flow_error = open_loop_flow_target - self.signed_flow.get::<gallon_per_second>();
+
+        let delta_error = flow_error - self.flow_error_prev.get::<gallon_per_second>();
+        self.flow_error_prev = VolumeRate::new::<gallon_per_second>(flow_error);
+
+        let p_term = Self::DEFAULT_P_GAIN * delta_error;
+        let i_term = Self::DEFAULT_I_GAIN * flow_error;
+
+        let force_gain = 200000.;
+        self.force += Force::new::<newton>((p_term + i_term) * force_gain);
+
+        if self.force > Force::new::<newton>(0.) {
+            if position_error > Ratio::new::<ratio>(0.)
+                && self.speed <= Velocity::new::<meter_per_second>(0.)
+            {
+                let max_force = hydraulic_pressure * self.bore_side_area;
+                self.force = self.force.min(max_force);
+            }
+        } else if position_error < Ratio::new::<ratio>(0.)
+            && self.speed >= Velocity::new::<meter_per_second>(0.)
+        {
+            let max_force = -1. * hydraulic_pressure * self.rod_side_area;
+            self.force = self.force.max(max_force);
+        }
+    }
+}
+
+impl Actuator for LinearActuator {
+    fn used_volume(&self) -> Volume {
+        self.volume_to_actuator_accumulator
+    }
+    fn reservoir_return(&self) -> Volume {
+        self.volume_to_res_accumulator
+    }
+    fn reset_accumulators(&mut self) {
+        self.volume_to_res_accumulator = Volume::new::<gallon>(0.);
+        self.volume_to_actuator_accumulator = Volume::new::<gallon>(0.);
+    }
+}
+
+pub trait HydraulicAssemblyController {
+    fn mode_requested(&self) -> LinearActuatorMode;
+    fn position_request(&self) -> Ratio;
+    fn should_lock(&self) -> bool;
+    fn lock_position_request(&self) -> Ratio;
+}
+pub struct HydraulicActuatorAssembly {
+    linear_actuator: LinearActuator,
+    rigid_body: RigidBodyOnHingeAxis,
```

Comment:
Fact is I think everything on the plane is door like :D 
I agree it's not super satisfying to have a particular actuator type in that structure. But after thinking about this there's pretty much no other type of actuator I would add there.  Well there is one indeed: the hydraulic motor actuators for flap and slat systems.  But the structure of flap/slat system is so different that I don't think I would use the same HydraulicActuatorAssembly container.  Maybe correct naming would be to have a HydraulicLinearActuatorAssembly and a HydraulicMotorActuatorAssembly ?   
Will have a second look at how this could be merged in a one only structure, but sounds difficult to me looking at how motors interact with flaps and slats. I'm still not too much at ease with the traits concept :p

Reply:
Ok, that's perfectly reasonable! Yeah, I agree with you that there's no point in trying to optimize prematurely here. We can think about making this more generic if we implement the flaps/slats with the hydraulic system and notice a lot of duplication. Only then is it worth considering merging some of the concepts. 👍